Contact Form: Add ability to set response status

tmoorewp opened the pull request:
Fixes #13001 

#### Changes proposed in this Pull Request:
Added functionality to parse feedback items into new buckets: "Responded", "In Progress", and "Needs Response". Parsing into each bucket can be done as a bulk action or individually on the action row of a feedback item.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is an update to the Contact Form module.

#### Testing instructions:
* You may need to deactivate/reactivate the Contact Form module (Settings -> Modules). New statuses are registered and WordPress may not pick them up otherwise.
* (optional) You'll need Feedback entries to test. Add them via a form.
* Go to the Feedback menu.
* Hover over a feedback item. You should see new "Responded", "In Progress", and "Needs A Response" links.
* Click a link. The Feedback item should be moved into its new bucket.
* Additionally, test adding multiple Feedback items. Check of two or more Feedback items, then select an action from the Bulk Edit menu.

#### Proposed changelog entry for your changes:
* Contact Form: Added the ability to mark feedbacks as "Responded", "In Progress", or "Needs A Response".

Commit:
Contact Form: Add ability to set response status (Responded, In Progress, or Needs Response).

See #13001

Commit:
Adding instructions for testers in test.md.

Commit:
- Fixed a typo that prevented bulk edits to "needs response."
- Updated the success message to be specific about the action taken, rather than always saying "marked as spam."

Commit:
Removing some whitespace.

jeherve reviewed (changes requested):
> You may need to deactivate/reactivate the Contact Form module (Settings -> Modules). New statuses are registered and WordPress may not pick them up otherwise.

Would there be a way to work around that? We can't really ask all current users to deactivate and reactivate the module.

*****

Would it be possible to add some unit tests to track that feature, to avoid any regressions in the future? The Contact Form includes so many different features, the more things we can cover with unit tests the better imo.

tmoorewp commented:
> > You may need to deactivate/reactivate the Contact Form module (Settings -> Modules). New statuses are registered and WordPress may not pick them up otherwise.
> 
> Would there be a way to work around that? We can't really ask all current users to deactivate and reactivate the module.

I added this line as a "just in case" for testing. Restarting the module should not be needed.

> Would it be possible to add some unit tests to track that feature, to avoid any regressions in the future? The Contact Form includes so many different features, the more things we can cover with unit tests the better imo.

I will work on some tests and add them ASAP.

tmoorewp commented:
@jeherve I'm using the Docker setup to test and develop Jetpack. I'm having trouble getting PHPUnit to run in this environment. I've read the docs in this repo, but using `yarn docker:phpunit` runs a lot of tests and doesn't seem to let me run individual tests (I've tried `yarn docker:phpunit tests/php/modules/contact-form/test-class.grunion-contact-form.php`, for example, and always get a `Cannot open file` error). Do you know of any guides to help with this?

jeherve commented:
What happens when you run `yarn docker:phpunit --filter WP_Test_Grunion_Contact_Form` instead? It should work.

If that does not help, try reinstalling the test suite with `yarn docker:update-core-unit-tests`.

Commit:
Adding a new check for unit testing. This is needed to cut off output of the HTML and exiting before the unit tests are done.

Commit:
1. Move the check of the DOING_JETPACK_UNIT_TEST constant.
2. Add unit tests for the grunion_ajax_spam function

Commit:
Fix some whitespace.

jeherve commented on modules/contact-form/admin.php:
```diff
@@ -918,6 +918,11 @@ function grunion_ajax_spam() {
 	}
 
+	// If unit tests are running, the output needs to be returned and not echoed.
+	if ( defined( 'DOING_JETPACK_UNIT_TEST' ) && DOING_JETPACK_UNIT_TEST ) {
```
You can use the existing `TESTING_IN_JETPACK` constant for scenarios like this one.

tmoorewp replied:
Thanks! I should've checked for this. Should be fixed now.

jeherve reviewed (changes requested):
Excellent, thank you! Here is a constant you could use in those cases.

Commit:
Use the existing testing constant, rather than making a new one.

Commit:
Use the existing testing constant rather than making a new one.

Commit:
Remove the definition of the testing constant.

Commit:
Unit tests for grunion_handle_bulk_spam function

Adding unit tests for the grunion_handle_bulk_spam function. Requires adding a check for the TESTING_IN_JETPACK to the function to get a return value.

Commit:
Adding some failure testing to grunion_ajax_spam

This makes sure that grunion_ajax_spam fails correctly if it doesn't get expected inputs.

tmoorewp commented:
@jeherve I've added testing for `grunion_ajax_spam` and `grunion_handle_bulk_spam`, which are the two functions changed for this feature enhancement.

jeherve commented:
All tests appear to be failing right now, so it may be worth taking a look at this:
https://travis-ci.org/Automattic/jetpack/builds/569369726?utm_source=github_status&utm_medium=notification

Commit:
Move changing of global $pagenow to setUp. Add a reset of $pagenow to tearDown().

Commit:
Clean up any variables we set during testing.

tmoorewp commented:
Sorry about that. Should be fixed now.

Commit:
Update the personal data checker and csv export functions to be aware of the new post statuses.

jeherve commented on modules/contact-form/admin.php:
```diff
@@ -784,6 +895,44 @@ function grunion_ajax_spam() {
 
+	if ( isset( $status['responded'] ) ) {
+		$status_html .= '<li><a href="edit.php?post_status=responded&amp;post_type=feedback"';
+		if ( $current_menu == 'responded' ) {
```
I would use Yoda conditions here as well.

```suggestion
		if ( 'responded' === $current_menu ) {
```

That said, and like above, I wonder if it may be worth refactoring this a bit so we don't have to keep repeating ourselves below.

jeherve commented on modules/contact-form/admin.php:
```diff
@@ -784,6 +895,44 @@ function grunion_ajax_spam() {
+
+	if ( isset( $status['inprogress'] ) ) {
+		$status_html .= '<li><a href="edit.php?post_status=inprogress&amp;post_type=feedback"';
+		if ( $current_menu == 'inprogress' ) {
```
```suggestion
		if ( 'inprogress' === $current_menu ) {
```

jeherve commented on modules/contact-form/admin.php:
```diff
@@ -784,6 +895,44 @@ function grunion_ajax_spam() {
+
+	if ( isset( $status['needsresponse'] ) ) {
+		$status_html .= '<li><a href="edit.php?post_status=needsresponse&amp;post_type=feedback"';
+		if ( $current_menu == 'needsresponse' ) {
```
```suggestion
		if ( 'needsresponse' === $current_menu ) {
```

jeherve commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -221,6 +221,43 @@ protected function __construct() {
+		// Add post statuses for responsded, inprogress, and needsresponse
+		register_post_status(
+			'responded', array(
+				'label'                  => 'Responded',
```
Should this be translated?

jeherve commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -221,6 +221,43 @@ protected function __construct() {
+
+		register_post_status(
+			'inprogress', array(
+				'label'                  => 'Response In Progress',
```
Should this be "In Progress" only, and should it be translatable?

jeherve commented:
Closing this for now because of the lack of activity on this. We can always reopen in the future if needed.
